README.md
# Pandoc-Suite and BibTeX-Zotero-Citation-Picker for Academic Writing in Markdown

---

⚠️ Since I changed my workflow substantially, this workflow will not be actively maintained anymore. The latest release still works and will most likely continue to work for the foreseeable future.

I am [continuing development of the citation picker part here](https://github.com/chrisgrieser/alfred-bibtex-citation-picker). For the pandoc conversion, I can recommend [DocDown](https://github.com/lowercasename/docdown) as an easy-to-use alternative.

---

![](https://img.shields.io/github/downloads/chrisgrieser/pandoc_alfred/total?label=Total%20Downloads&style=plastic)  ![](https://img.shields.io/github/v/release/chrisgrieser/pandoc_alfred?label=Latest%20Release&style=plastic)

## What it does
This Alfred workflow enables academic citing and publishing with any markdown app. It does so by offering a **Citation Picker** for [Pandoc Markdown Citations](https://pandoc.org/MANUAL.html#pandocs-markdown) and by easily **converting Markdown files** into `.docx`, `.pdf`, `.html`, `.odt`, or `.pptx` with the proper bibliography.

This workflow is made for academics who neither like Word nor LaTeX. As the name suggests, this workflow is basically a user-friendly interface for Pandoc, allowing you to make use of Pandoc's capabilities without ever having to open the Terminal.

The workflow otherwise also features some useful auxiliary functions, like for example "anticipatory word count", i.e.  number of words that your bibliography will have after the conversion (markdown apps cannot calculate the word count of the bibliography as it hasn't been created yet).

## Download
➡️ [Latest Release on GitHub](https://github.com/chrisgrieser/pandoc_alfred/releases)

➡️ [Complete Usage & Installation Guide](https://chris-grieser.de/pandoc_alfred)

## Screenshots

<img src="https://i.imgur.com/hBkN8e3.png" alt="Conversion" width=50% height=50%>

<img src="https://i.imgur.com/XuSfGov.png" alt="Zotero Citation Picker" width=50% height=50%>

<img width=50% height=50% alt="Screenshot 2021-09-09 22 08 42" src="https://user-images.githubusercontent.com/73286100/132755578-cce9892e-d3c0-4ba3-9666-4649d8b96202.png">

## Credits

### Donations
Are much appreciated via [PayPal](https://www.paypal.com/paypalme/ChrisGrieser) or [Ko-Fi](https://ko-fi.com/pseudometa) 🙏

### About the Author
This workflow has been created by [@pseudo_meta (Twitter)](https://twitter.com/pseudo_meta) aka Chris Grieser (rl). In my day job, I am a PhD student in sociology, studying the governance of the app economy. If you are interested in this subject, check out [my academic homepage](https://chris-grieser.de/) and get in touch.


Changelog.md
- 2022-08-20	release 5.3.6
- 2022-08-20	remove mail contact since workflow unmaintained
- 2022-02-14	local csl folder not needed as value anymore
- 2022-02-08	retiring anticipatory word count feature
- 2022-01-31	release 5.3.5
- 2022-01-31	fix bibtex title key parsing for alternative formatting (#8)
- 2022-01-18	release 5.3.4
- 2022-01-18	release 5.3.3
- 2022-01-18	release 5.3.2
- 2021-10-31	Update FUNDING.yml
- 2021-10-05	fix #4
- 2021-09-18	5.2
- 2021-09-05	Create FUNDING.yml
- 2021-08-05	Merge pull request #3 from manuelhubacher/main
- 2021-08-05	feat: Use date field
- 2021-06-02	added `2>&1` in the command construction to enable error-log saving later in the workflow
- 2021-05-09	info.plist
- 2021-05-08	info.plist
- 2021-05-08	info.plist
- 2021-04-26	v. 1.2.1
- 2021-04-26	v. 1.2.1
- 2020-11-16	version 1.2
- 2020-10-22	Create README.md
- 2020-10-22	Initial commit


scripts/buffer_writer.js
#!/usr/bin/env osascript -l JavaScript
function run() {
	ObjC.import("stdlib");
	const app = Application.currentApplication();
	app.includeStandardAdditions = true;

	const alfredBarLength = parseInt ($.getenv("alfred_bar_length"));
	const urlIcon = $.getenv("IconURL");
	const doiIcon = $.getenv("IconDOI");
	const library = $.getenv("bibtex_library_path").replace(/^~/, app.pathTo("home folder"));
	const library2 = $.getenv("second_library").replace(/^~/, app.pathTo("home folder"));

	// read bib file &  remove unnecessary info to increase speed
	let input = app.doShellScript("cat \"" + library + "\" | grep -vwE \"(abstract|annotate|Bdsk-Url-1|Bdsk-Url-2|date-modified|date-added|keywords|issn|langid|urlyear|isbn|location|pagetotal|series|eprint) =\" | grep -vw \"%%\"");
	if (library2) input += "\r" + app.doShellScript("cat \"" + library2 + "\" | grep -vwE \"(abstract|annotate|Bdsk-Url-1|Bdsk-Url-2|date-modified|date-added|keywords|issn|langid|urlyear|isbn|location|pagetotal|series|eprint) =\" | grep -vw \"%%\"");

	// BibTeX-Decoding
	const germanChars = ["{\\\"u};ü", "{\\\"a};ä", "{\\\"o};ö", "{\\\"U};Ü", "{\\\"A};Ä", "{\\\"O};Ö", "\\\"u;ü", "\\\"a;ä", "\\\"o;ö", "\\\"U;Ü", "\\\"A;Ä", "\\\"O;Ö", "\\ss;ß", "{\\ss};ß"];
	const otherChars = ["{\\~n};ñ", "{\\'a};á", "{\\'e};é", "{\\v c};č", "\\c{c};ç", "\\o{};ø", "\\^{i};î", "\\\"{i};î", "\\\"{i};ï", "{\\'c};ć", "\\\"e;ë"];
	const specialChars = ["\\&;&", "``;\"", "`;'", "\\textendash{};—", "---;—", "--;—"];
	const decodePair = [...germanChars, ...otherChars, ...specialChars];
	decodePair.forEach(pair => {
		const half = pair.split(";");
		input = input.replaceAll (half[0], half[1]);
	});

	const inputArray = input.split("@");
	const entryArray = [];

	// extracts content of a BibTeX-field
	function extract (str) {
		str = str.split(" = ")[1];
		return str.replace (/,$/, "");
	}

	inputArray.forEach(entry => {
		const properties = entry.split ("\r");
		let author = "";
		let title = "";
		let year = "";
		let editor = "";
		let url = "";
		let collection = "";
		let doi = "";
		let volume = "";
		let issue = "";
		let numberOfEditors = 0;
		const citekey = properties[0].replace (/.*{(.*),/, "$1");
		const type = properties[0].replace (/(.*){.*/, "$1");

		properties.forEach (property => {
			property = property.replace (/[{|}]/g, ""); // remove Tex

			if (property.includes ("author =")) {
				author = extract(property)
					.replace (/(, [A-Z]).+?(?= and|$)/gm, "") // remove first names
					.replaceAll (" and ", " & ")
					.replace (/&.*&.*/, "et al."); // insert et al
			}
			else if (property.includes ("editor =")) {
				editor = extract(property)
					.replace (/(, [A-Z]).+?(?= and|$)/gm, "") // remove first names
					.replaceAll (" and ", " & ");
				numberOfEditors = editor.split("&").length;
				editor = editor.replace (/&.*&.*/, "et al."); // insert et al
			}
			else if (/title =/i.test(property)) {
				title = extract(property);
				if (title.length > alfredBarLength) title = title.substring(0, alfredBarLength).trim() + "...";
			}
			else if (property.includes ("year =")) year = property.replace (/.*=\s*{?(\d{4}).*/, "$1");
			else if (property.includes ("date =")) year = property.replace (/.*=\s*{?(\d{4}).*/, "$1");
			else if (property.includes ("doi =")) doi = extract(property);
			else if (property.includes ("url =")) url = extract(property);
			else if (property.includes ("volume =")) volume = extract(property);
			else if (property.includes ("number =")) issue = "(" + extract(property) + ")";
			else if (/\s(journal|booktitle)\s*=/i.test(property)) collection = "    In: " + extract(property);

		});

		// if the initial reading of the file cut away the closing bracket, add it
		let bibtexEntry = "@" + entry;
		const openingBrackets = (bibtexEntry.match(/\{/g) || []).length;
		const closingBrackets = (bibtexEntry.match(/\}/g) || []).length;
		if (openingBrackets > closingBrackets) bibtexEntry = bibtexEntry + "}";

		// when no URL, try to use DOI
		let urlAppendix = "";
		let URLsubtitle = "⛔️ There is no URL or DOI.";
		if (url) {
			URLsubtitle = "⌃: Open URL " + urlIcon;
			urlAppendix = "    " + urlIcon;
			if (doi) urlAppendix = urlAppendix + " " + doiIcon;
		} else if (doi) {
			URLsubtitle = "⌃: Open DOI " + doiIcon;
			urlAppendix = "    " + doiIcon;
		}

		// icon selection
		let typeIcon = "icons/";
		switch (type) {
			case "article":
				typeIcon += "article.png";
				collection = collection + " " + volume + issue;
				break;
			case "incollection":
			case "inbook":
				typeIcon += "book_chapter.png";
				break;
			case "misc":
			case "unpublished":
				typeIcon += "manuscript.png";
				break;
			case "book":
				typeIcon += "book.png";
				break;
			case "techreport":
				typeIcon += "technical_report.png";
				break;
			case "inproceedings":
				typeIcon += "conference.png";
				break;
			case "online":
			case "webpage":
				typeIcon += "website.png";
				break;
		}

		// determines correct editor-abbreviation
		let editorAbbrev = "(Ed.)";
		if (numberOfEditors > 1) editorAbbrev = "(Eds.)";

		// displays editor when there are no authors
		let authoreditor = author + " ";
		if (!author && editor) authoreditor = editor + " " + editorAbbrev + " ";
		else if (!author && !editor) authoreditor = "";

		const alfredMatcher = [title, author, editor, year, collection, citekey]
			.join(" ")
			.replaceAll ("-", " ");

		entryArray.push ({
			"title": title,
			"autocomplete": authoreditor,
			"subtitle": authoreditor + year + collection + urlAppendix,
			"match": alfredMatcher,
			"arg": "@" + citekey,
			"icon": { "path": typeIcon },
			"uid": citekey,
			"text": { "copy": url },
			"mods": {
				"ctrl": {
					"valid": (url !== ""),
					"arg": url,
					"subtitle": URLsubtitle,
				},
				"fn": { "arg": bibtexEntry }
			}
		});
	});

	return JSON.stringify({ "items": entryArray });
}


scripts/pconf.sh
#!/bin/zsh
cat << EOB
{
	"items": [
		{
			"title": "🚪 Exit this Configuration",
			"match":"exit leave quit cancel",
			"uid": "1",
			"arg": "exit"
		},
		{
			"title": "❓ Documentation & Guide for the Pandoc-Suite",
			"arg": "workflow_documentation",
			"match":"help documentation usage guide",
			"uid": "2",
			"subtitle": "Troubleshooting & further Information"
		},
		{
			"title": "❓ Pandoc Documentation",
			"arg": "man_pandoc",
			"match":"help documentation usage guide manual",
			"uid": "3",
			"subtitle": "open man page in Preview.app"
		},
		{
			"title": "🆕 ✍️ Search & Select Citation Style",
			"arg": "csl_file",
			"match":"citation style csl",
			"uid": "4",
			"subtitle": "currently: $csl_file"
		},
		{
			"title": "📚 1️⃣ BibTeX Library",
			"arg": "bibtex_library_path",
			"match":"bib bibtex library references",
			"uid": "5",
			"subtitle": "currently: $bibtex_library_path"
		},
		{
			"title": "📚 🤿 Debugging: Deep Search Library File",
			"arg": "hard_bib_search",
			"match":"bib bibtex library references",
			"uid": "6",
			"subtitle": "Select this if your BibTex Library cannot be found with the option above"
		},
		{
			"title": "📇 Set Resource Path subfolder",
			"match":"resource path ressource images attachments",
			"arg": "resource_path_subfolder",
			"uid": "7",
			"subtitle": "currently: $resource_path_subfolder"
		},
		{
			"title": "🔘 Hotkey Output Format",
			"match":"hotkey output format docx html pdf pptx",
			"arg": "hotkey_output_format",
			"uid": "8",
			"subtitle": "currently: $hotkey_output_format"
		},
		{
			"title": "📅 Date to Add (filename & content)",
			"match":"date format filename add insert american",
			"arg": "date_to_append",
			"uid": "9",
			"subtitle": "currently: $date_to_append"
		},
		{
			"title": "📄 Word Reference Document",
			"match":"word docx reference document template office microsoft",
			"arg": "reference_docx_path",
			"uid": "10",
			"subtitle": "currently: $reference_docx_path"
		},
		{
			"title": "🏎 PDF-Engine",
			"match":"pdf-engine pdf engine pdflatex wkhtmltopdf groff",
			"arg": "pdf_engine",
			"uid": "11",
			"subtitle": "currently: $pdf_engine"
		},
		{
			"title": "🧑‍🏫 Powerpoint Reference Presentation",
			"match":"reference powerpoint presentation pptx template office microsoft slides",
			"arg": "reference_pptx_path",
			"uid": "12",
			"subtitle": "currently: $reference_pptx_path"
		},
		{
			"title": "📊 Slide Level for Powerpoint",
			"match":"slide level for powerpoint presentation pptx",
			"arg": "slide_level",
			"uid": "13",
			"subtitle": "currently: $slide_level"
		},
		{
			"title": "✴️ Citation Picker",
			"match":"zotero citation picker alfred classic alternative",
			"arg": "citation_picker",
			"uid": "14",
			"subtitle": "currently: $citation_picker"
		},
		{
			"title": "📚 2️⃣ Additional Library (optional)",
			"arg": "second_library",
			"match":"additional second library bib bibtex references",
			"uid": "15",
			"subtitle": "currently: $second_library"
		},
		{
			"title": "❇️ Further Arguments",
			"match":"additional further more arguments options",
			"arg": "further_pandoc_args",
			"uid": "16",
			"subtitle": "currently: $further_pandoc_args"
		},
		{
			"title": "⚙️ Writer Extensions",
			"match":"extensions writer additional options",
			"arg": "writer_extensions",
			"uid": "17",
			"subtitle": "currently: $writer_extensions"
		},
		{
			"title": "⚙️ Reader Extensions",
			"match":"extensions reader additional options",
			"arg": "reader_extensions",
			"uid": "18",
			"subtitle": "currently: $reader_extensions"
		},
		{
			"title": "🚰 Pandoc Filter (used before --citeproc)",
			"match":"filter addtional options",
			"arg": "pandoc_filter",
			"uid": "19",
			"subtitle": "currently: $pandoc_filter"
		},
		{
			"title": "📄 Pandoc Template",
			"match":"tempaltes addtional options",
			"arg": "pandoc_template",
			"uid": "20",
			"subtitle": "currently: $pandoc_template"
		},
		{
			"title": "🔄 Debugging: Force Citation Picker Buffer Reload",
			"match":"debugging buffer citation picker reload cache reset bug force",
			"arg": "force_buffer_reload",
			"uid": "21",
			"subtitle": "Try this when the Alfred Citation Picker misbehaves"
		},
		{
			"title": "🚀 Open Output File after Conversion",
			"match":"open output after conversion Finder reveal",
			"arg": "open_after",
			"uid": "23",
			"subtitle": "currently: $open_after"
		},
		{
			"title": "📄 OpenOffice Reference Document",
			"match":"writer openoffice office odt reference document template",
			"arg": "reference_odt_path",
			"uid": "24",
			"subtitle": "currently: $reference_odt"
		},
		{
			"title": "🔗 Link Citations to Bibliography & Bibliography to URLs",
			"match":"link autolink URL hyperlink citations bibliography",
			"arg": "link_citations_biblio",
			"uid": "25",
			"subtitle": "currently: $link_citations_biblio"
		}
	]
}

EOB
